src/pybind/mgr/cephadm/agent.py

```diff
@@ -0,0 +1,218 @@
+import cherrypy
+import json
+import socket
+import tempfile
+import threading
+import time
+
+from mgr_util import create_self_signed_cert, verify_tls_files
+from ceph.utils import datetime_now
+
+from typing import Set, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from cephadm.module import CephadmOrchestrator
+
+
+class CherryPyThread(threading.Thread):
+    def __init__(self, mgr: "CephadmOrchestrator") -> None:
+        self.mgr = mgr
+        self.cherrypy_shutdown_event = threading.Event()
+        super(CherryPyThread, self).__init__(target=self.run)
+
+    def run(self) -> None:
+        class Root(object):
+            exposed = True
+
+            def __init__(self, mgr: "CephadmOrchestrator"):
+                self.mgr = mgr
+
+            def GET(self) -> str:
+                return '''<!DOCTYPE html>
+<html>
+    <head><title>Cephadm HTTP Endpoint</title></head>
+    <body>
+        <p>Cephadm HTTP Endpoint is up and running</p>
+    </body>
+</html>'''
+
+            def POST(self, host: str, ls: str = '', networks: str = '', facts: str = '', ack: str = '', keyring: str = '') -> None:
+                try:
+                    handler = RequestHandler(
+                        self.mgr, host, ls, networks, facts, ack, keyring)
+                    handler.start()
+                except Exception as e:
+                    self.mgr.log.warning(
+                        f'Cephadm http endpoint failed to handle http POST request: {e}')
+
+        server_addr = self.mgr.get_mgr_ip()
+        server_port = self.mgr.endpoint_port
+
+        cert, key = create_self_signed_cert(common_name='cephadm-agent')
+
+        self.key_tmp = tempfile.NamedTemporaryFile()
+        self.key_tmp.write(key.encode('utf-8'))
+        self.key_tmp.flush()  # pkey_tmp must not be gc'ed
+        key_fname = self.key_tmp.name
+
+        self.cert_tmp = tempfile.NamedTemporaryFile()
+        self.cert_tmp.write(cert.encode('utf-8'))
+        self.cert_tmp.flush()  # cert_tmp must not be gc'ed
+        cert_fname = self.cert_tmp.name
+
+        verify_tls_files(cert_fname, key_fname)
```

Comment:
We could additionally provide the cert to the client so that it does a client-side verification too

Reply:
+1 we shouldn't be using verify=false. The cert build will need to be more generic though (IIRC, I did something similar in the exporter code)

Reply:
Done. We now generate a root cert and a cert signed by that to have the cephadm endpoint use for ssl. The root cert is given to the agents when deployed in order to verify the ssl cert the cephadm endpoint is using.